model/githubapp/github_app_installation.go

```diff
@@ -286,11 +289,40 @@ func (c *installationTokenCache) get(installationID int64, lifetime time.Duratio
 // installation token can be used before it expires.
 const MaxInstallationTokenLifetime = time.Hour
 
-func (c *installationTokenCache) put(installationID int64, installationToken string, createdAt time.Time) {
+func (c *installationTokenCache) put(installationID int64, installationToken string, permissions *github.InstallationPermissions, createdAt time.Time) {
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.cache[installationID] = cachedInstallationToken{
+	id := createCacheID(installationID, permissions)
+	c.cache[id] = cachedInstallationToken{
 		installationToken: installationToken,
 		expiresAt:         createdAt.Add(MaxInstallationTokenLifetime),
 	}
 }
+
+// createCacheID creates an ID based on the installation ID and the token's permissions.
+// This allows us to put and get installation tokens from the cache based on the installation ID
+// and the permissions that the token is scoped to.
+// The format of the ID is: "<installationID>_<permissionKey-permissionValue>_<permissionKey-permissionValue>...".
+func createCacheID(installationID int64, permissions *github.InstallationPermissions) string {
+	id := fmt.Sprint(installationID)
+	if permissions == nil {
+		return id
+	}
+	val := reflect.ValueOf(permissions).Elem()
```

Comment:
nit: variable name would be a little clearer if it was something like `permissionsStructVal` or something along those lines.